stan/math/prim/prob/beta_lpdf.hpp

```diff
@@ -60,122 +65,79 @@ return_type_t<T_y, T_scale_succ, T_scale_fail> beta_lpdf(
-        digamma_beta[n] = digamma(beta_dbl);
-      }
-    }
+    logp -= sum(lgamma(beta_val)) * N / max_size(beta);
   }
-
-  VectorBuilder<include_summand<propto, T_scale_succ, T_scale_fail>::value,
-                T_partials_return, T_scale_succ, T_scale_fail>
-      lgamma_alpha_beta(max_size(alpha, beta));
-  VectorBuilder<!is_constant_all<T_scale_succ, T_scale_fail>::value,
-                T_partials_return, T_scale_succ, T_scale_fail>
-      digamma_alpha_beta(max_size(alpha, beta));
-
-  if (include_summand<propto, T_scale_succ, T_scale_fail>::value) {
-    for (size_t n = 0; n < max_size(alpha, beta); n++) {
-      const T_partials_return alpha_beta
-          = value_of(alpha_vec[n]) + value_of(beta_vec[n]);
-      lgamma_alpha_beta[n] = lgamma(alpha_beta);
-      if (!is_constant_all<T_scale_succ, T_scale_fail>::value) {
-        digamma_alpha_beta[n] = digamma(alpha_beta);
-      }
-    }
+  if (include_summand<propto, T_y, T_scale_succ>::value) {
+    logp += sum((alpha_val - 1.0) * log_y) * N / max_size(y, alpha);
+  }
+  if (include_summand<propto, T_y, T_scale_fail>::value) {
+    logp += sum((beta_val - 1.0) * log1m_y) * N / max_size(y, beta);
   }
 
-  for (size_t n = 0; n < N; n++) {
-    const T_partials_return y_dbl = value_of(y_vec[n]);
-    const T_partials_return alpha_dbl = value_of(alpha_vec[n]);
-    const T_partials_return beta_dbl = value_of(beta_vec[n]);
-
-    if (include_summand<propto, T_scale_succ, T_scale_fail>::value) {
-      logp += lgamma_alpha_beta[n];
-    }
-    if (include_summand<propto, T_scale_succ>::value) {
-      logp -= lgamma_alpha[n];
-    }
-    if (include_summand<propto, T_scale_fail>::value) {
-      logp -= lgamma_beta[n];
-    }
-    if (include_summand<propto, T_y, T_scale_succ>::value) {
-      logp += (alpha_dbl - 1.0) * log_y[n];
-    }
-    if (include_summand<propto, T_y, T_scale_fail>::value) {
-      logp += (beta_dbl - 1.0) * log1m_y[n];
+  if (!is_constant_all<T_y>::value) {
+    if (is_vector<T_y>::value) {
+      ops_partials.edge1_.partials_ = forward_as<T_partials_matrix>(
+          (alpha_val - 1) / y_val + (beta_val - 1) / (y_val - 1));
+    } else {
+      ops_partials.edge1_.partials_[0]
+          = sum((alpha_val - 1) / y_val + (beta_val - 1) / (y_val - 1));
     }
+  }
 
-    if (!is_constant_all<T_y>::value) {
-      ops_partials.edge1_.partials_[n]
-          += (alpha_dbl - 1) / y_dbl + (beta_dbl - 1) / (y_dbl - 1);
-    }
-    if (!is_constant_all<T_scale_succ>::value) {
-      ops_partials.edge2_.partials_[n]
-          += log_y[n] + digamma_alpha_beta[n] - digamma_alpha[n];
-    }
-    if (!is_constant_all<T_scale_fail>::value) {
-      ops_partials.edge3_.partials_[n]
-          += log1m_y[n] + digamma_alpha_beta[n] - digamma_beta[n];
+  if (include_summand<propto, T_scale_succ, T_scale_fail>::value) {
+    const auto& alpha_beta
+        = to_ref_if<!is_constant_all<T_scale_succ, T_scale_fail>::value>(
+            alpha_val + beta_val);
```

Comment:
This ```if``` clause has the gradients only being calculated if ```include_summand<propto, T_scale_succ, T_scale_fail>::value``` is true, which is different from the existing header. Is this intentional?

Reply:
It may look different, but it is actually not. `include_summand<propto, T_scale_succ, T_scale_fail>::value` is true if either `propto=false` or any of the `T_scale_succ, T_scale_fail` is non-constant. So if gradients need to be calculated this condition will always be true.